google-ads/src/main/java/com/google/ads/googleads/lib/utils/ResourceNames.java

```diff
@@ -0,0 +1,191 @@
+import com.google.ads.googleads.v0.resources.AdGroupName;
+import com.google.ads.googleads.v0.resources.BiddingStrategyName;
+import com.google.ads.googleads.v0.resources.BillingSetupName;
+import com.google.ads.googleads.v0.resources.CampaignBidModifierName;
+import com.google.ads.googleads.v0.resources.CampaignBudgetName;
+import com.google.ads.googleads.v0.resources.CampaignCriteriaName;
+import com.google.ads.googleads.v0.resources.CampaignGroupName;
+import com.google.ads.googleads.v0.resources.CampaignName;
+import com.google.ads.googleads.v0.resources.CampaignSharedSetName;
+import com.google.ads.googleads.v0.resources.ChangeStatusName;
+import com.google.ads.googleads.v0.resources.ConversionActionName;
+import com.google.ads.googleads.v0.resources.CustomerName;
+import com.google.ads.googleads.v0.resources.GeoTargetConstantName;
+import com.google.ads.googleads.v0.resources.GoogleAdsFieldName;
+import com.google.ads.googleads.v0.resources.KeywordViewName;
+import com.google.ads.googleads.v0.resources.RecommendationName;
+import com.google.ads.googleads.v0.resources.SharedCriteriaName;
+import com.google.ads.googleads.v0.resources.SharedSetName;
+import com.google.ads.googleads.v0.resources.VideoName;
+import com.google.common.base.Joiner;
+import java.util.Arrays;
+
+/**
+ * Utilities for generating resource names. Offers several advantages over the utilities in the
+ * various {@code *Name} classes.
+ *
+ * <ul>
+ *   <li>Method parameter types match the type of each identifier. For example, the {@code
+ *       campaignId} parameter of {@link #campaign(long, long)} is of type {@code long}. In
+ *       contrast, the {@code campaign} parameter of {@link CampaignName#format(String, String)} is
+ *       of type {@code String}.
+ *   <li>Method parameter lists contain each component identifier. For example, {@link
+ *       #adGroupCriterion(long, long, long)} expects a customer ID, an ad group ID, and a criterion
+ *       ID. In contrast, {@link AdGroupCriteriaName#format(String, String)} expects a customer ID
+ *       and an underscore-delimited combination of ad group ID and criterion ID.
+ * </ul>
+ */
+public class ResourceNames {
+
+  /** Joiner for resource path components that consist of multiple identifiers. */
+  private static final Joiner IDENTIFIER_JOINER = Joiner.on('_');
+
+  /**
+   * Returns a string consisting of the specified identifiers, concatenated together using the
+   * standard delimiter for composite identifiers.
+   */
+  private static String concatLongs(long... components) {
+    return IDENTIFIER_JOINER.join(Arrays.stream(components).mapToObj(Long::toString).iterator());
+  }
+
+  /** Returns the account budget proposal resource name for the specified components. */
+  public static String accountBudgetProposal(long customerId, long accountBudgetProposalId) {
+    return AccountBudgetProposalName.format(
+        Long.toString(customerId), concatLongs(accountBudgetProposalId));
```

Comment:
Why do you occasionally use "concatLongs" on individual IDs? Should these be Long.toString as well?

Reply:
Good catch. I've updated all of the single component cases to just use Long.toString.